snapcraft/internal/build_providers/_base_provider.py

```diff
@@ -0,0 +1,97 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import abc
+import shlex
+from typing import Callable
+
+import petname
+
+
+class Provider():
+
+    def __init__(self, *, project, echoer) -> None:
+        self.project = project
+        self.echoer = echoer
+        # Once https://github.com/CanonicalLtd/multipass/issues/220 is
+        # closed we can prepend snapcraft- again.
+        self.instance_name = petname.Generate(2, '-')
+        self.project_dir = shlex.quote(project.info.name)
+
+        if project.info.version:
+            self.snap_filename = '{}_{}_{}.snap'.format(
+                project.info.name, project.info.version, project.deb_arch)
+        else:
+            self.snap_filename = '{}_{}.snap'.format(
+                project.info.name, project.deb_arch)
+
+    def __enter__(self):
+        self.create()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.destroy()
+
+    @property
+    @abc.abstractmethod
+    def run(self) -> Callable[[str], None]:
```

Comment:
It seems odd to have this as a property, especially considering that it's actually a callable, so it LOOKS like a method when used. Why not just make it a method, and the implementer can just pass it on to their command instance?